### ISSUE TITLE
Rule proposal: `prefer-logical-nullish-assignment`

### ISSUE THREAD

**Issue by boris-petrov** [eslint/eslint#14721]
I searched and didn't see something like that requested...

**Please describe what the rule should do:**

```ts
if (x == null) {
  x = 1;
}
```

Could be rewritten as:

```ts
x ??= 1;
```

**What new ECMAScript feature does this rule relate to?**

[Logical nullish assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_nullish_assignment).

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```ts
if (x.y == null) {
  x.y = 1;
}
```

```ts
const object = {};
if (!('x' in object)) {
  object.x = [];
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Because it improves JS code by using core JS features.

**Are you willing to submit a pull request to implement this rule?**

No.

**Comment by nzakas** [comment]
I think this may be too narrow a use case for a core rule. I could see maybe adding something that is a general `prefer-nullish-coalescing` that would catch assignments as well as expressions, but I don't think we'd want to just flag assignments.

**Comment by boris-petrov** [comment]
Well, there already is a [prefer-nullish-coalescing](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-nullish-coalescing.md) (albeit in `typescript-eslint`) but it doesn't work for assignment (it's even named in a way that would suggest so). That's why I opened the issue for the other case - assignment.

**Comment by ljharb** [comment]
Since there's now `&&=`, `||=`, and `??=`, what about a rule that detected all the cases that could use compound assignment?

In other words:
```jsx
x = x + n;
x = x - n;
x = x && n;
x = x || n;
x = x ?? n;

if (x) {
  x = n;
}
if (!x) {
  x = n;
}
if (x != null) {
  x = n;
}
```
would all transform (or vice versa) to:
```jsx
x += n;
x -= n;
x &&= n;
x ||- n;
x ??= n;

x &&= n;
x ||= n;
x ??= n;
```

**Comment by btmills** [comment]
We have [`operator-shorthand`](https://eslint.org/docs/rules/operator-assignment) for the mathematical operators. Would #13689 cover the logical operator request here?

**Comment by glen-84** [comment]
This exact syntax is mentioned by @mdjermanovic [here](https://github.com/eslint/eslint/issues/13689#issuecomment-692465982). I think that it can be covered by that rule.

**Comment by boris-petrov** [comment]
I just noticed another case where that should work:

```javascript
const object = {};
if (!('x' in object)) {
  object.x = [];
}
```

Could be rewritten as:

```javascript
const object = {};
object.x ??= [];
```

**Comment by ljharb** [comment]
Those have slightly different semantics, though - if Object.prototype.x existed as undefined or null, the first one wouldn’t set it, but the second would.

**Comment by boris-petrov** [comment]
Yes, you're right. Maybe that shouldn't be handled in that case...

**Comment by btmills** [comment]
Closing to combine into #13689. If you've left a :+1: here, feel free to add one there as well.
